contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+
+```golang
+// +nonNamespaced=true
+// +genclient=true
+
+type ResourceClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        // Spec defines resources required
+        Spec ResourceClassSpec
+        // +optional
+        Status ResourceClassStatus
+}
+// Spec defines resources required
+type ResourceClassSpec struct {
+        // Resource Selector selects resources
+        ResourceSelector []ResourcePropertySelector
+}
+
+// A null or empty selector matches no resources
+type ResourcePropertySelector struct {
+        // A list of resource/device selector requirements. ANDed from each ResourceSelectorRequirement
+        MatchExpressions []ResourceSelectorRequirement
+}
+
+// A resource selector requirement is a selector that contains values, a key, and an operator
+// that relates the key and values
+type ResourceSelectorRequirement struct {
+        // The label key that the selector applies to
+        // +patchMergeKey=key
+        // +patchStrategy=merge
+        Key string
+        // +optional
+        Values []string
+        // operator
+        Operator ResourceSelectorOperator
+}
+type ResourceSelectorOperator string
+
+const (
+        ResourceSelectorOpIn           ResourceSelectorOperator = "In"
+        ResourceSelectorOpEq           ResourceSelectorOperator = "Eq"
+        ResourceSelectorOpNotIn        ResourceSelectorOperator = "NotIn"
+        ResourceSelectorOpExists       ResourceSelectorOperator = "Exists"
+        ResourceSelectorOpDoesNotExist ResourceSelectorOperator = "DoesNotExist"
+        ResourceSelectorOpGt           ResourceSelectorOperator = "Gt"
+        ResourceSelectorOpGtEq         ResourceSelectorOperator = "GtEq"
+        ResourceSelectorOpLt           ResourceSelectorOperator = "Lt"
+        ResourceSelectorOpLtEq         ResourceSelectorOperator = "LtEq"
+)
+```
+### ResourceClassStatus
+```golang
+type ResourceClassStatus struct {
+        Allocatable resources.Quantity
+        Requested   resources.Quantity
+}
+```
+ResourceClass status is updated by the scheduler at:
+1. New *Resource Class* object creation.
+2. Node addition to the cluster.
+3. Node removal from the cluster.
+4. Pod creation if pod requests a resource class.
+5. Pod deletion if pod was consuming resource class.
+
+`ResourceClassStatus` serves the following two purposes:
+* Scheduler predicates evaluation while pod creation. For details, please refer
+  further sections
+* User can view the current usage/availability details about the resource class
+  using kubectl.
+
+### User story
+The administrator has deployed device plugins to support hardware present in the
+cluster. Device plugins, running on nodes, will update node status indicating
+the presence of this hardware. To offer this hardware to applications deployed
+on kubernetes in a portable way, the administrator creates a number of resource
+classes to represent that hardware. These resource classes will include metadata
```

Comment:
> the administrator creates a number of resource classes to represent that hardware

Hi, I am just wondering why we create more than one resource class to represent one hardware, should it be one by one matching? 

In other word, one hardware need and only need one reource class to represent it (with multi selectors of course).

Please let me know if i missing something about this proposal, thanks~

Reply:
Let say there are two NICs: NIC1 40 GB and NIC2 40GB + vlan offloading
Now two classes can be created:
1. Class offering only 40G speed NICs
2. Class offering vlan offloading capable NICs.

You see both the classes can select NIC1. This is what it meant by `number of resource classes to represent ...` . Does that make sense?

PS: sorry for late response.

Reply:
Thanks for your explanation.

Reply:
A deviceinfo structure will be maintained in schedulercache. So each device will have a list of what all resource class it is satisfying.
Take a look: https://github.com/vikaschoudhary16/kubernetes/pull/3/files#diff-558cb8bde14dca10a3151bfc222a3aae

Reply:
This was actually my point above about separating quantitative and qualitative things. If users were able to ask for an amount of 1 of resource class of `nic`, an amount of 40 (or >= 40) of resource class `nic.speed_bits_per_second` and a separate qualitative string of `nic.offload.vlan`, then device plugins can advertise resource inventories (amounts) separately from capabilities (like offloads). As this proposal stands right now, you will end up having to create quite a few resource classes, with each of the resource classes varying only slightly based on capabilities.